### ISSUE TITLE
Bio SwissProt.read() broken in Python 3.5 (biopython 1.66 np110py35_0)

### ISSUE THREAD

**Issue by fpmenninger** [biopython/biopython#773]
Windows7 Python 3.5.1 Anaconda 2.5.0 biopython 1.66 np110py35_0
(in jupyter)
```
from Bio import ExPASy    
handle = ExPASy.get_sprot_raw("P75952")
from Bio import SwissProt
record = SwissProt.read(handle)
```

results in:
```
---------------------------------------------------------------------------
IncompleteRead                            Traceback (most recent call last)
<ipython-input-1-11c8db89238e> in <module>()
      3 print(handle)
      4 from Bio import SwissProt
----> 5 record = SwissProt.read(handle)
      6 print(record)

C:\Anaconda3\lib\site-packages\Bio\SwissProt\__init__.py in read(handle)
    135     ##print(handle.url," ",handle," ",handle.fp)
    136     ##handle.fp = None
--> 137     remainder = handle.read()
    138     print(remainder)
    139     if remainder:

C:\Anaconda3\lib\http\client.py in read(self, amt)
    446             else:
    447                 try:
--> 448                     s = self._safe_read(self.length)
    449                 except IncompleteRead:
    450                     self._close_conn()

C:\Anaconda3\lib\http\client.py in _safe_read(self, amt)
    596             print(chunk, "\n")
    597             if not chunk:
--> 598                 raise IncompleteRead(b''.join(s), amt)
    599             s.append(chunk)
    600             amt -= len(chunk)

IncompleteRead: IncompleteRead(0 bytes read, 6066 more expected)
```

from what I can tell (not much experience here)

it seems something changed in how lines are read in `__init__.py` of SwissProt
such that when handle.read() of http/client is called,
different values are sent than were before.

The Python 3.5.1 version (biopython 1.66 np110py35_0) passes http/client.read a header with 
length = 6066 and fp = <_io.BufferedReader name=1156>
whereas Python 3.4.4 (biopython 1.66 np110py34_0) version passes a header with
length = 0 and fp = None

**Comment by peterjc** [comment]
Is this a bytes vs unicode thing? Would try debugging this by first downloading the data to a file, then parsing the file.

**Comment by peterjc** [comment]
Example works with Python 3.4.1 on the Mac; don't have Python 3.5 on this laptop yet...

**Comment by fpmenninger** [comment]
Using
```
from Bio import ExPASy    
handle = ExPASy.get_sprot_raw("P75952")
print(handle)
```
gives a handle of "http://www.uniprot.org/uniprot/P75952.txt "
Saved this file to "P75952.dat" and SwissProt.read() parsed correctly.
Don't have enough experience here to answer 'bytes v unicode', so don't know next step in debugging, but thanks for the help.  Apparently one workaround is to save the files locally and parse from there.

**Comment by peterjc** [comment]
Good news: I could reproduce this on Mac OS X using Python 3.5.0, using the latest Biopython from GitHub (i.e. 1.66+)

```
$ python3.5
Python 3.5.0 (default, Sep 14 2015, 12:13:24) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Bio import ExPASy    
>>> handle = ExPASy.get_sprot_raw("P75952")
>>> from Bio import SwissProt
>>> record = SwissProt.read(handle)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Volumes/DataDrive/repositories/biopython/Bio/SwissProt/__init__.py", line 134, in read
    remainder = handle.read()
  File "/Users/pjcock/lib/python3.5/http/client.py", line 446, in read
    s = self._safe_read(self.length)
  File "/Users/pjcock/lib/python3.5/http/client.py", line 594, in _safe_read
    raise IncompleteRead(b''.join(s), amt)
http.client.IncompleteRead: IncompleteRead(0 bytes read, 6066 more expected)
>>> quit()
```

Or, calling ``urlopen`` directly:

```
$ python3.5
Python 3.5.0 (default, Sep 14 2015, 12:13:24) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib.request import urlopen
>>> handle = urlopen("http://www.uniprot.org/uniprot/P75952.txt")
>>> from Bio import SwissProt
>>> record = SwissProt.read(handle)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Volumes/DataDrive/repositories/biopython/Bio/SwissProt/__init__.py", line 134, in read
    remainder = handle.read()
  File "/Users/pjcock/lib/python3.5/http/client.py", line 446, in read
    s = self._safe_read(self.length)
  File "/Users/pjcock/lib/python3.5/http/client.py", line 594, in _safe_read
    raise IncompleteRead(b''.join(s), amt)
http.client.IncompleteRead: IncompleteRead(0 bytes read, 6066 more expected)
>>> quit()
```

Both of those example work fine on the same machine using Python 3.4.

Interestingly the error happens at line 134, ``remainder = handle.read()``, at which point we should be at the end of the file (if it was a single record file), and expect to get back an empty string.

Something has subtly changed in Python 3.5 (perhaps a bug)...

**Comment by fpmenninger** [comment]
Again, thank you.
So, not a problem lies in python 3.5, not in biopython?
Inserting some print statements in `http/client.py` and `__init__.py` of SwissProt.read shows how the old version reads one character at a time, reducing the header.length() by one with each read.
The new version somehow reads complete lines and does not alter the header.length().
How is this now dealt with?  Is there someone monitoring these issues that might submit a bug report to python?  Or is that for me to do?

**Comment by peterjc** [comment]
Right now I do suspect a bug in Python 3.5 itself. I'd want to reduce this down to a simple self contained test not using Biopython at all, and then file this at http://bugs.python.org/ (checking it isn't already logged), and make a note of the issue here.

If you are interested in doing this detective work instead of me, that would be great :+1:

**Comment by fpmenninger** [comment]
I'd be happy to do the detective work, but fear I've not the in depth skill set to know how to set up such a test.  I did try to see if the bug was logged, but again ran into the difficulty of being a bit clueless as to how I would reference the bug.  
If you are willing, please point me towards where I can begin sleuthing... (I think it lies in something that happens in http.client.py read() function) and how I would reference the bug when looking to see if it was logged.

**Comment by peterjc** [comment]
Progress, I was able to make a self-contained Python example, apparently showing something subtle has changed about how Python handles leaving the handle's iterator context:

```python
import sys
from urllib.request import urlopen

print(sys.version)

def count_till_slashes(bytes_handle):
    count = 0
    for line in bytes_handle:
        count += 1
        if line.startswith(b"//"):
            break  # Or, can use: return count
    return count

handle = urlopen("http://www.uniprot.org/uniprot/P75952.txt")
count = count_till_slashes(handle)
print("URL had %i interesting lines" % count)
remainder = handle.read()
print("Then %i bytes we don't care about" % len(remainder))
```

Calling that script ``issue773.py`` testing under 64 bit Linux CentOS 6 gives:

```
$ python3.2 issue773.py
3.2 (r32:88445, Nov  7 2012, 22:05:23) 
[GCC 4.4.6 20120305 (Red Hat 4.4.6-4)]
URL had 127 interesting lines
Then 0 bytes we don't care about

$ python3.3 issue773.py
3.3.0 (default, Nov  7 2012, 21:52:39) 
[GCC 4.4.6 20120305 (Red Hat 4.4.6-4)]
URL had 127 interesting lines
Then 0 bytes we don't care about

$ python3.4 issue773.py
3.4.3 (default, Aug 21 2015, 11:12:32) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-11)]
URL had 127 interesting lines
Then 0 bytes we don't care about

$ python3.5 issue773.py
3.5.0 (default, Sep 28 2015, 11:25:31) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-16)]
URL had 127 interesting lines
Traceback (most recent call last):
  File "issue773.py", line 17, in <module>
    remainder = handle.read()
  File "/home/pc40583/lib/python3.5/http/client.py", line 446, in read
    s = self._safe_read(self.length)
  File "/home/pc40583/lib/python3.5/http/client.py", line 594, in _safe_read
    raise IncompleteRead(b''.join(s), amt)
http.client.IncompleteRead: IncompleteRead(0 bytes read, 6066 more expected)
```

Time to read up on the Python issue tracker etc to see what might be going on.

**Comment by peterjc** [comment]
Filed with the Python team as http://bugs.python.org/issue26499

Leaving this issue open, we might have to add a workaround to Biopython.

**Comment by peterjc** [comment]
It looks like a tiny change at ``Bio/SwissProt/__init__.py`` line 134 from ``remainder = handle.read()`` to ``remainder = handle.readline()`` is enough to avoid the Python 3.5 problem reported as http://bugs.python.org/issue26499

I've made this change in Biopython as https://github.com/biopython/biopython/commit/dbb0de6337d5604ea5a5b2276dfb0ae0d2ddd1dc so you could either get the latest code from GitHub and re-install from source, wait for the next release, or just edit that one line by hand to get the fix.

Thank you again for reporting this @fpmenninger - this not only helped Biopython, but the wider Python community as well.

**Comment by peterjc** [comment]
Good news: http://bugs.python.org/issue26499 has been fixed, and should be included in the next Python 3.5.x release (and in Python 3.6 too).

**Comment by wayoung** [comment]
Has this been fixed in 3.5.1? 'Cause I am still getting this problem.

**Comment by peterjc** [comment]
According to https://docs.python.org/3/whatsnew/changelog.html Python issue 26499 is fixed in Python 3.5.2 rc 1 (i.e. it will be fixed in Python 3.5.2, but was not fixed in Python 3.5.1)
